### ISSUE TITLE
"Custom Mapping" Display Value option is not available on tables with NULL entries 

### ISSUE THREAD

**Issue by shawa** [metabase/metabase#8116]
Expected behaviour: Custom value mappings should be available for all columns
Actual behaviour: Mappings only available for columns containing no null values

When choosing display values for Enum members, the "Custom Values" option only appears if there are no `null` entries in the column in question, even though Metabase can enumerate the members of the column for say, dropdowns in field filters.

Steps to reproduce:
1. Create a table with a nullable column and populate with values, some null
2. Navigate to `Admin` > `Data Model` > `[Table]` > `[nullable column]` > `Settings gear`
3. Display Values
    - Only _Use original value_ is available
4. Remove any rows with `nulls` from the table
    - _Custom Values_ is now available

I'm not overly familiar with the codebase, nor React, but this appears to be the [line in question](
https://github.com/metabase/metabase/blob/c6ea68f0cb32c328baa5019645f87e955ca3c1a1/frontend/src/metabase/admin/datamodel/containers/FieldApp.jsx#L523).

| Info  | Value |
| ------------- | ------------- |
| Your Browser and the version  | Google Chrome 67.0.3396.99 (Official Build) (64-bit)  |
| Your operating system| OS X 10.13.5 |
| Your databases | Postgres|
| Metabase version|  v0.29.3 |
| Metabase hosting environment| Docker under Elastic Beanstalk |
| Metabase internal database| Postgres|

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by zedidas** [comment]
We are implementing Metabase on a big database with a lost of historical data. Some of the fields did not exist previously so there are empty.

Due to the issue explained in this ticket, we are not able to use Custom Value Mapping in a lot of our columns. The work around appears to be to review the past data to eliminate any remaining null value but we would prefer to avoid changing any historical data.

Can we expect that this may be fixed in a future release? Is there any work around that would allow us to use custom mapping without having to change the data?

**Comment by zedidas** [comment]
Can someone provide an update on this. IMHO, it appears to be a fairly simple solution to implement, and that would extend the functionality available.

I would greatly appreciate some info on this. And also some info about any existing work-around. 

Thanks!

**Comment by salsakran** [comment]
When you remove null values does that change the field type?

Also, can you share the schema for that table and what field type Metabase guesses for that column?

**Comment by zedidas** [comment]
So I have made a very simple test. Here's a table I've created on MS SQL Express:

```
CREATE TABLE Products (
    id int,
    name varchar(255),
    type int,
    subtype int,
    PRIMARY KEY (id)
);

INSERT INTO Products VALUES (1,'Car', 1, 1);
INSERT INTO Products VALUES (2,'Plane', 1, 2);
INSERT INTO Products VALUES (3,'Bike', 1, 1);
INSERT INTO Products VALUES (4,'Boat', 1, NULL);
INSERT INTO Products VALUES (5,'Milk', 2, 4);
INSERT INTO Products VALUES (6,'Water', 2, 5);
```

As you can see, I am using two int columns for type and subtype. Column subtype will have a single entry which is NULL.

After I configure the database in the admin page, this is how the table looks:
![image](https://user-images.githubusercontent.com/5529181/47788303-33e6ad80-dd09-11e8-8f48-dd1d141660f1.png)

Now if I try to configure Custom Mapping on the column Type, here's what I see:
![image](https://user-images.githubusercontent.com/5529181/47788337-57a9f380-dd09-11e8-93b0-33977da8ec88.png)

If I try to do the same for the column Type, here's what I see:
![image](https://user-images.githubusercontent.com/5529181/47788368-6c868700-dd09-11e8-98db-c74564fdd576.png)

As you can see, although both columns types are the same (int), since the column subtype has a NULL value, the Custom Mapping feature is not shown for that column.

This is a test I did: As soon as I update that NULL value as an int, the Custom Mapping option becomes available.

In my humble opinion, it would be interesting to allow custom mapping when the column contains NULL values. One could specify the Custom Mapping to apply to the NULL value itself.

Let me know if I can help with additional info.

**Comment by zedidas** [comment]
Hello,

Any chance this is going to be addressed in a future release? I am detecting more and more situations where the custom mapping is not possible only because there are a few entries with NULL values.

Thanks!

**Comment by urkle** [comment]
We actually have a column where null is a legitimate value. (it's a pay period and null means off, vs a number means it's set to a certain period.. weekly, monthly, etc..). And I've been fussing w/ Metabase for the past 15 minutes trying to figure out why it wouldn't let me define custom mapping!

**Comment by maxmalysh** [comment]
It happened to me even when there were no NULL entries in my table. I had to "Discard cached field values". After reloading the admin page, I got "Custom mapping" as an option.

Postgres 11 + Metabase v0.32.2.

**Comment by mnort9** [comment]
Is this a client-side issue? If so, I can take a stab at it. I'm not too familiar with Clojure though, so not much I can do on the backend side.

This is the one issue holding us back from fully migrating to Metabase. We use a decent amount of enums with valid NULL values.

**Comment by brandonlipman** [comment]
Any updates on this issue or suggested fixes for the short term? @mazameli

**Comment by uhrohraggy** [comment]
Running into this as well. Nullable fields are legitimate for lots of reasons, would love not needing to update historical values or converting to SQL for every query and writing this as a `CASE  ... WHEN  ... ELSE ... END`

**Comment by rjurado01** [comment]
This error occurs with all columns, even when there are no `null` values.

**Comment by mazameli** [comment]
Bumping up priority on this given how many folks have run into it.

**Comment by michaelflux** [comment]
Bumping this one as remains an issue.

**Comment by camsaul** [comment]
It's not clear at this point whether this is a frontend issue or a backend issue. It's been a while so I don't recall how we decide which options get shown on that page. I don't recall whether they're determined entirely client-side or if some of it is based on what comes back from the API calls. If someone could help triage by figuring out where these options are determined and whether there's a difference in relevant API responses for the columns with the option and the ones without then it would make tackling this issue pretty easy for us and we can get a fix out.

**Comment by paulrosenzweig** [comment]
Fixed in #13033
